// @inworld/nodejs-sdk/build/src/entities/packets/latency/perceived_latency_report.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerceivedLatencyReport = exports.PerceivedLatencyReportPrecisionType = void 0;
const packets_pb_1 = require("../../../../../proto/ai/inworld/packets/packets_pb");
var PerceivedLatencyReportPrecisionType;
(function (PerceivedLatencyReportPrecisionType) {
    PerceivedLatencyReportPrecisionType["UNSPECIFIED"] = "UNSPECIFIED";
    PerceivedLatencyReportPrecisionType["FINE"] = "FINE";
    PerceivedLatencyReportPrecisionType["ESTIMATED"] = "ESTIMATED";
    PerceivedLatencyReportPrecisionType["PUSH_TO_TALK"] = "PUSH_TO_TALK";
    PerceivedLatencyReportPrecisionType["NON_SPEECH"] = "NON_SPEECH";
})(PerceivedLatencyReportPrecisionType || (exports.PerceivedLatencyReportPrecisionType = PerceivedLatencyReportPrecisionType = {}));
class PerceivedLatencyReport {
    constructor({ latency, precision, }) {
        this.latency = latency;
        this.precision = precision;
    }
    static fromProto(proto) {
        return new PerceivedLatencyReport({
            latency: proto.getLatency(),
            precision: this.getPerceivedLatencyReportPrecision(proto.getPrecision()),
        });
    }
    static getPerceivedLatencyReportPrecision(precision) {
        switch (precision) {
            case packets_pb_1.PerceivedLatencyReport.Precision.FINE:
                return PerceivedLatencyReportPrecisionType.FINE;
            case packets_pb_1.PerceivedLatencyReport.Precision.ESTIMATED:
                return PerceivedLatencyReportPrecisionType.ESTIMATED;
            case packets_pb_1.PerceivedLatencyReport.Precision.PUSH_TO_TALK:
                return PerceivedLatencyReportPrecisionType.PUSH_TO_TALK;
            case packets_pb_1.PerceivedLatencyReport.Precision.NON_SPEECH:
                return PerceivedLatencyReportPrecisionType.NON_SPEECH;
            default:
                return PerceivedLatencyReportPrecisionType.UNSPECIFIED;
        }
    }
    static getProtoPerceivedLatencyReportPrecision(precision) {
        switch (precision) {
            case PerceivedLatencyReportPrecisionType.FINE:
                return packets_pb_1.PerceivedLatencyReport.Precision.FINE;
            case PerceivedLatencyReportPrecisionType.ESTIMATED:
                return packets_pb_1.PerceivedLatencyReport.Precision.ESTIMATED;
            case PerceivedLatencyReportPrecisionType.PUSH_TO_TALK:
                return packets_pb_1.PerceivedLatencyReport.Precision.PUSH_TO_TALK;
            case PerceivedLatencyReportPrecisionType.NON_SPEECH:
                return packets_pb_1.PerceivedLatencyReport.Precision.NON_SPEECH;
            default:
                return packets_pb_1.PerceivedLatencyReport.Precision.UNSPECIFIED;
        }
    }
}
exports.PerceivedLatencyReport = PerceivedLatencyReport;
