// @inworld/nodejs-sdk/build/src/entities/entities/entity_item.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityItem = void 0;
const entities_packets_pb_1 = require("../../../../proto/ai/inworld/packets/entities/entities_packets_pb");
class EntityItem {
    constructor({ id, displayName, description, properties }) {
        this.properties = {};
        this.id = id;
        if (displayName) {
            this.displayName = displayName;
        }
        if (description) {
            this.description = description;
        }
        if (Object.keys(properties).length > 0) {
            this.properties = properties;
        }
    }
    static fromProto(proto) {
        return new EntityItem({
            id: proto.getId(),
            displayName: proto.getDisplayName(),
            description: proto.getDescription(),
            properties: Object.fromEntries(proto.getPropertiesMap().entries()),
        });
    }
    toProto() {
        const proto = new entities_packets_pb_1.EntityItem()
            .setId(this.id)
            .setDisplayName(this.displayName)
            .setDescription(this.description);
        const properties = proto.getPropertiesMap();
        for (const [key, value] of Object.entries(this.properties)) {
            properties.set(key, value);
        }
        return proto;
    }
}
exports.EntityItem = EntityItem;
