// @inworld/nodejs-sdk/build/src/entities/entities/item_operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ItemOperation = void 0;
const entities_packets_pb_1 = require("../../../../proto/ai/inworld/packets/entities/entities_packets_pb");
const data_structures_1 = require("../../common/data_structures");
const entity_item_1 = require("./entity_item");
class ItemOperation {
    constructor({ createOrUpdateItems, removeItems, itemsInEntities, }) {
        if (createOrUpdateItems) {
            this.createOrUpdateItems = createOrUpdateItems;
        }
        if (removeItems) {
            this.removeItems = removeItems;
        }
        if (itemsInEntities) {
            this.itemsInEntities = itemsInEntities;
        }
    }
    static fromProto(proto) {
        var _a, _b, _c, _d;
        const itemsInEntities = proto.getItemsInEntities();
        const type = itemsInEntities === null || itemsInEntities === void 0 ? void 0 : itemsInEntities.getType();
        const ids = (_a = proto.getRemoveItems()) === null || _a === void 0 ? void 0 : _a.getItemIdsList();
        return new ItemOperation(Object.assign(Object.assign({ createOrUpdateItems: {
                items: (_c = (_b = proto
                    .getCreateOrUpdateItems()) === null || _b === void 0 ? void 0 : _b.getItemsList()) === null || _c === void 0 ? void 0 : _c.map((item) => entity_item_1.EntityItem.fromProto(item)),
                addToEntities: (_d = proto.getCreateOrUpdateItems()) === null || _d === void 0 ? void 0 : _d.getAddToEntitiesList(),
            } }, (ids && { removeItems: { itemIds: ids } })), { itemsInEntities: Object.assign(Object.assign({}, (type && { type: this.convertOperationTypeFromProto(type) })), { itemIds: itemsInEntities === null || itemsInEntities === void 0 ? void 0 : itemsInEntities.getItemIdsList(), entityNames: itemsInEntities === null || itemsInEntities === void 0 ? void 0 : itemsInEntities.getEntityNamesList() }) }));
    }
    toProto() {
        var _a;
        if (this.createOrUpdateItems) {
            return new entities_packets_pb_1.ItemsOperationEvent().setCreateOrUpdateItems(new entities_packets_pb_1.CreateOrUpdateItemsOperation()
                .setAddToEntitiesList(this.createOrUpdateItems.addToEntities)
                .setItemsList(this.createOrUpdateItems.items.map((item) => item.toProto())));
        }
        if (this.removeItems) {
            return new entities_packets_pb_1.ItemsOperationEvent().setRemoveItems(new entities_packets_pb_1.RemoveItemsOperation().setItemIdsList(this.removeItems.itemIds));
        }
        if ((_a = this.itemsInEntities) === null || _a === void 0 ? void 0 : _a.type) {
            return new entities_packets_pb_1.ItemsOperationEvent().setItemsInEntities(new entities_packets_pb_1.ItemsInEntitiesOperation()
                .setType(ItemOperation.convertOperationTypeToProto(this.itemsInEntities.type))
                .setItemIdsList(this.itemsInEntities.itemIds)
                .setEntityNamesList(this.itemsInEntities.entityNames));
        }
        else {
            return new entities_packets_pb_1.ItemsOperationEvent();
        }
    }
    static convertOperationTypeToProto(type) {
        switch (type) {
            case data_structures_1.ItemsInEntitiesOperationType.ADD:
                return entities_packets_pb_1.ItemsInEntitiesOperation.Type.ADD;
            case data_structures_1.ItemsInEntitiesOperationType.REMOVE:
                return entities_packets_pb_1.ItemsInEntitiesOperation.Type.REMOVE;
            case data_structures_1.ItemsInEntitiesOperationType.REPLACE:
                return entities_packets_pb_1.ItemsInEntitiesOperation.Type.REPLACE;
        }
    }
    static convertOperationTypeFromProto(type) {
        switch (type) {
            case entities_packets_pb_1.ItemsInEntitiesOperation.Type.ADD:
                return data_structures_1.ItemsInEntitiesOperationType.ADD;
            case entities_packets_pb_1.ItemsInEntitiesOperation.Type.REMOVE:
                return data_structures_1.ItemsInEntitiesOperationType.REMOVE;
            case entities_packets_pb_1.ItemsInEntitiesOperation.Type.REPLACE:
                return data_structures_1.ItemsInEntitiesOperationType.REPLACE;
        }
    }
}
exports.ItemOperation = ItemOperation;
