// @inworld/nodejs-sdk/build/src/common/data_structures/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogLevel = exports.ItemsInEntitiesOperationType = exports.ConversationParticipant = exports.UnderstandingMode = exports.MicrophoneMode = exports.ConversationIntializeState = exports.InworldLatencyReportType = exports.InworldConversationEventType = exports.InworlControlAction = exports.InworldTextPacketType = exports.InworldPacketType = exports.ConnectionState = exports.ConversationState = void 0;
var ConversationState;
(function (ConversationState) {
    ConversationState["ACTIVE"] = "ACTIVE";
    ConversationState["PROCESSING"] = "PROCESSING";
    ConversationState["INACTIVE"] = "INACTIVE";
})(ConversationState || (exports.ConversationState = ConversationState = {}));
var ConnectionState;
(function (ConnectionState) {
    ConnectionState["ACTIVE"] = "ACTIVE";
    ConnectionState["ACTIVATING"] = "ACTIVATING";
    ConnectionState["INACTIVE"] = "INACTIVE";
})(ConnectionState || (exports.ConnectionState = ConnectionState = {}));
var InworldPacketType;
(function (InworldPacketType) {
    InworldPacketType["UNKNOWN"] = "UNKNOWN";
    InworldPacketType["TEXT"] = "TEXT";
    InworldPacketType["AUDIO"] = "AUDIO";
    InworldPacketType["TASK"] = "TASK";
    InworldPacketType["LOG"] = "LOG";
    InworldPacketType["TRIGGER"] = "TRIGGER";
    InworldPacketType["EMOTION"] = "EMOTION";
    InworldPacketType["CONTROL"] = "CONTROL";
    InworldPacketType["SILENCE"] = "SILENCE";
    InworldPacketType["CANCEL_RESPONSE"] = "CANCEL_RESPONSE";
    InworldPacketType["NARRATED_ACTION"] = "NARRATED_ACTION";
    InworldPacketType["SCENE_MUTATION_REQUEST"] = "SCENE_MUTATION_REQUEST";
    InworldPacketType["SCENE_MUTATION_RESPONSE"] = "SCENE_MUTATION_RESPONSE";
    InworldPacketType["CONVERSATION_UPDATE"] = "CONVERSATION_UPDATE";
    InworldPacketType["CONVERSATION_EVENT"] = "CONVERSATION_EVENT";
    InworldPacketType["ENTITIES_ITEM_OPERATION"] = "ENTITIES_ITEM_OPERATION";
    InworldPacketType["OPERATION_STATUS"] = "OPERATION_STATUS";
    InworldPacketType["LATENCY_REPORT"] = "LATENCY_REPORT";
})(InworldPacketType || (exports.InworldPacketType = InworldPacketType = {}));
var InworldTextPacketType;
(function (InworldTextPacketType) {
    InworldTextPacketType["UNKNOWN"] = "UNKNOWN";
    InworldTextPacketType["SPEECH_TO_TEXT"] = "SPEECH_TO_TEXT";
    InworldTextPacketType["TYPED_IN"] = "TYPED_IN";
    InworldTextPacketType["GENERATED"] = "GENERATED";
    InworldTextPacketType["FILLER"] = "FILLER";
})(InworldTextPacketType || (exports.InworldTextPacketType = InworldTextPacketType = {}));
var InworlControlAction;
(function (InworlControlAction) {
    InworlControlAction["UNKNOWN"] = "UNKNOWN";
    InworlControlAction["AUDIO_SESSION_START"] = "AUDIO_SESSION_START";
    InworlControlAction["AUDIO_SESSION_END"] = "AUDIO_SESSION_END";
    InworlControlAction["CONVERSATION_EVENT"] = "CONVERSATION_EVENT";
    InworlControlAction["CONVERSATION_UPDATE"] = "CONVERSATION_UPDATE";
    InworlControlAction["INTERACTION_END"] = "INTERACTION_END";
    InworlControlAction["TTS_PLAYBACK_MUTE"] = "TTS_PLAYBACK_MUTE";
    InworlControlAction["TTS_PLAYBACK_UNMUTE"] = "TTS_PLAYBACK_UNMUTE";
    InworlControlAction["WARNING"] = "WARNING";
})(InworlControlAction || (exports.InworlControlAction = InworlControlAction = {}));
var InworldConversationEventType;
(function (InworldConversationEventType) {
    InworldConversationEventType["UNKNOWN"] = "UNKNOWN";
    InworldConversationEventType["STARTED"] = "STARTED";
    InworldConversationEventType["UPDATED"] = "UPDATED";
    InworldConversationEventType["EVICTED"] = "EVICTED";
})(InworldConversationEventType || (exports.InworldConversationEventType = InworldConversationEventType = {}));
var InworldLatencyReportType;
(function (InworldLatencyReportType) {
    InworldLatencyReportType["PERCEIVED_LATENCY"] = "PERCEIVED_LATENCY";
    InworldLatencyReportType["PING_PONG"] = "PING_PONG";
})(InworldLatencyReportType || (exports.InworldLatencyReportType = InworldLatencyReportType = {}));
var ConversationIntializeState;
(function (ConversationIntializeState) {
    ConversationIntializeState["ACTIVE"] = "ACTIVE";
    ConversationIntializeState["PROCESSING"] = "PROCESSING";
    ConversationIntializeState["INACTIVE"] = "INACTIVE";
})(ConversationIntializeState || (exports.ConversationIntializeState = ConversationIntializeState = {}));
var MicrophoneMode;
(function (MicrophoneMode) {
    MicrophoneMode["OPEN_MIC"] = "OPEN_MIC";
    MicrophoneMode["EXPECT_AUDIO_END"] = "EXPECT_AUDIO_END";
})(MicrophoneMode || (exports.MicrophoneMode = MicrophoneMode = {}));
var UnderstandingMode;
(function (UnderstandingMode) {
    UnderstandingMode["FULL"] = "FULL";
    UnderstandingMode["SPEECH_RECOGNITION_ONLY"] = "SPEECH_RECOGNITION_ONLY";
})(UnderstandingMode || (exports.UnderstandingMode = UnderstandingMode = {}));
var ConversationParticipant;
(function (ConversationParticipant) {
    ConversationParticipant["USER"] = "USER";
})(ConversationParticipant || (exports.ConversationParticipant = ConversationParticipant = {}));
var ItemsInEntitiesOperationType;
(function (ItemsInEntitiesOperationType) {
    ItemsInEntitiesOperationType["ADD"] = "ADD";
    ItemsInEntitiesOperationType["REMOVE"] = "REMOVE";
    ItemsInEntitiesOperationType["REPLACE"] = "REPLACE";
})(ItemsInEntitiesOperationType || (exports.ItemsInEntitiesOperationType = ItemsInEntitiesOperationType = {}));
var LogLevel;
(function (LogLevel) {
    LogLevel["UNSPECIFIED"] = "UNSPECIFIED";
    LogLevel["WARNING"] = "WARNING";
    LogLevel["INFO"] = "INFO";
    LogLevel["DEBUG"] = "DEBUG";
    LogLevel["INTERNAL"] = "INTERNAL";
})(LogLevel || (exports.LogLevel = LogLevel = {}));
