// @inworld/nodejs-sdk/build/src/entities/character.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Character = void 0;
class Character {
    constructor(props) {
        this.id = props.id;
        this.resourceName = props.resourceName;
        this.displayName = props.displayName;
        this.assets = props.assets;
    }
    static fromProto(proto) {
        const assets = proto.getCharacterAssets();
        return new Character({
            id: proto.getAgentId(),
            resourceName: proto.getBrainName(),
            displayName: proto.getGivenName(),
            assets: {
                avatarImg: assets === null || assets === void 0 ? void 0 : assets.getAvatarImg(),
                avatarImgOriginal: assets === null || assets === void 0 ? void 0 : assets.getAvatarImgOriginal(),
                rpmModelUri: assets === null || assets === void 0 ? void 0 : assets.getRpmModelUri(),
                rpmImageUriPortrait: assets === null || assets === void 0 ? void 0 : assets.getRpmImageUriPortrait(),
                rpmImageUriPosture: assets === null || assets === void 0 ? void 0 : assets.getRpmImageUriPosture(),
            },
        });
    }
}
exports.Character = Character;
