// @inworld/nodejs-sdk/build/src/entities/scene.entity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scene = void 0;
const character_entity_1 = require("./character.entity");
class Scene {
    constructor(props) {
        var _a, _b, _c;
        this.name = props.name;
        this.characters = (_a = props.characters) !== null && _a !== void 0 ? _a : [];
        this.description = (_b = props.description) !== null && _b !== void 0 ? _b : '';
        this.displayName = (_c = props.displayName) !== null && _c !== void 0 ? _c : '';
    }
    static serialize(scene) {
        return JSON.stringify(scene);
    }
    static deserialize(json) {
        try {
            const { name, characters } = JSON.parse(json);
            return new Scene({ name, characters });
        }
        catch (e) { }
    }
    static fromProto(proto) {
        return new Scene({
            name: proto.getSceneName(),
            characters: proto
                .getAgentsList()
                .map((agent) => character_entity_1.Character.fromProto(agent)),
            description: proto.getSceneDescription(),
            displayName: proto.getSceneDisplayName(),
        });
    }
    getCharactersByIds(ids) {
        return this.characters.filter((c) => ids.includes(c.id));
    }
    getCharactersByResourceNames(names) {
        return this.characters.filter((c) => names.includes(c.resourceName));
    }
}
exports.Scene = Scene;
